### ISSUE TITLE
Missing compiler error for `@extern` function with unspecified calling convention

### ISSUE THREAD

**Issue by Rexicon226** [ziglang/zig#19086]
## Steps

A bit involved but I've found consistent.

1. Create a file `main.zig`:
```zig
const std = @import("std");

pub fn main() !void {
    var ptr: *u8 = undefined; // Can be anything. "var ptr: *u8 = &x;" fails in the same way.

    print(@intFromPtr(&ptr));
    print(@intFromPtr(&ptr));

    _ = &ptr;
}

const maybePrint = @extern(?*const fn (usize) void, .{ .name = "print_function" });
fn print(addr: usize) void {
    maybePrint.?(addr);
}
```

2. Create a file `lib.zig`:
```zig
const std = @import("std");

comptime {
    @export(print, .{ .name = "print_function", .linkage = .Strong });
}

fn print(addr: usize) callconv(.C) void {
    std.debug.print("Addr: {}\n", .{addr});
}
```

3. Compile
Here are the commands you need to run. **NOTE: All of these are assumed to be x86_64 backend. Pass in `-fno-llvm` if needed.**
- `zig build-obj lib.zig`
- `zig build-exe main.zig lib.o`

4. Run

Output will show that the addresses do not match. Example:
![image](https://github.com/ziglang/zig/assets/87927264/88b7a586-7443-4182-9e7e-2a720dca2a1a)

Here is the exact same steps, just compiled with LLVM:
![image](https://github.com/ziglang/zig/assets/87927264/34c98bb0-c3fd-4a79-a99c-7f4f99e29762)

Discovered while writing a MemorySanitizer to be used in Zig (hopefully :pray:). Was trying to track addresses, and was finding they weren't matching when on x86_64 which was causing missed UB detection.

**Comment by nektro** [comment]
if that's the case then calling `@extern` with a `callconv(.Unspecified)` should've been a compile error

**Comment by Rexicon226** [comment]
> Your definition of maybePrint in the first file is incorrect: you don't specify the calling convention. Adding callconv(.C) to the type makes the program behave as expected.

I see mlugg, thanks. I should have noticed that :)

**Comment by Rexicon226** [comment]
fixed landed in: 51706af908e0c6acb822ef36760b7fe31faf62a6
